[FEAT] TurboModules guide (#3168)
facebook/react-native-website: docs/the-new-architecture/pillars-turbomodule.md

- Strongly typed interfaces that are consistent across platforms
- The ability to create shared C++ code for use across platforms
- Lazy loading of modules, allowing for faster app startup
- The use of JSI, a JavaScript interface for native code, which allows for more efficient communication between native and JavaScript code than the bridge

This guide will show you how to create a basic TurboModule.

:::caution
TurboModules only work with the **New Architecture** enabled.
To migrate to the **New Architecture**, follow the [Migration guide](../new-architecture-intro)
:::

## How to Create a TurboModule

To create a TurboModule, we need to:

1. Define the JavaScript specification.
2. Configure the module so that CodeGen can generate the scaffolding.
3. Write the native code to finish implementing the module.

## 1. Folder Setup

In order to keep the module decoupled from the app, it's a good idea to define the module separately from the app, and then add it as a dependency to your app later. This is also what you'll do for writing TurboModules that can be released as open-source libraries later.

> Next to your application, create a folder called `RTNCalculator`. (**RTN** stands for "**R**eact**T** **N**ative", and is a standard prefix for React Native modules).

Author: I think "standard" is also okay to mean what you're conveying here ("a standard" can mean "a required method of doing thing", but it can also mean "conventional"). But "conventional prefix" is also okay, but it sounds more awkward to me.

> Next to your application, create a folder called `RTNCalculator`. (**RTN** stands for "**R**eact**T** **N**ative", and is a standard prefix for React Native modules).

Author: Addressed, chose "recommended" as I think it conveys the idea better

<Tabs groupId="turbomodule-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
<TabItem value="flow">

```typescript title="NativeCalculator.js"
// @flow
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}
export default (TurboModuleRegistry.get<Spec>(
  'RTNCalculator'
): ?Spec);
```

</TabItem>
<TabItem value="typescript">

```typescript title="NativeCalculator.ts"
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}

export default (TurboModuleRegistry.get<Spec>(
  'RTNCalculator'
) as Spec | null);
```

</TabItem>
</Tabs>

At the beginning of the spec files are the imports:

- The `TurboModule` type, which defines the base interface for all TurboModules
- The `TurboModuleRegistry` JavaScript module, which contains functions for loading TurboModules

> The second section of the file contains the interface specification for the TurboModule. In this case, the interface defines the `add` function which takes two numbers and returns a promise that resolves to a number. This interface type **must** be named `Spec` for a TurboModule.

Reviewer: > This interface type **must** be named `Spec` for a TurboModule.

Not sure about this bit. I'll make some test: we are defining an interface and the `TurboModuleRegistry.get` method is generic over some type... so... we should be able to pick the name we prefer, theoretically...

Author: I tried to generate it without this name and it didn't work, I think the error message even told me it needed to be `Spec`.

  "react-native": "js/index",
  "source": "js/index",
  "files": [
    "js",
    "android",
    "ios",
    "rtn-calculator.podspec",
    "!android/build",
    "!ios/build",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__"
  ],
  "keywords": [
    "react-native",
    "ios",
    "android"
  ],
  "repository": "https://github.com/<your_github_handle>/rtn-calculator",
  "author": "<Your Name> <your_email@your_provider.com> (https://github.com/<your_github_handle>)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/<your_github_handle>/rtn-calculator/issues"
  },
  "homepage": "https://github.com/<your_github_handle>/rtn-calculator#readme",
  "devDependencies": {},
  "peerDependencies": {
    "react": "*",
    "react-native": "*"
  },
  "codegenConfig": {
    "libraries": [
      {
        "name": "RTNCalculatorSpec",
        "type": "modules",
        "jsSrcsDir": "js"
      }
    ]
  }
}
```

### iOS: Create the `podspec` file

For iOS, you'll need to create a `rtn-calculator.podspec` file which will define the module as a dependency for your app. It will stay in the root of `RTNCalculator`, alongside the `ios` folder.

The file will look like this:

> ```ruby title="rnt-calculator.podspec"

Reviewer: ```suggestion
```ruby title="rtn-calculator.podspec"
```

require "json"

package = JSON.parse(File.read(File.join(__dir__, "package.json")))

> folly_version = '2021.07.22.00'

Reviewer: React native 69 still comes with `2021.06.28.00-v2`. Not sure whether we want to use the new version: this guide will not work using `2021.07.22`...

Reviewer 2: Yes my bad :( You're actually right. I forgot that this guide is not based upon the current status of `main`. Apologize for the back and forth

Author: Addressed

:::

TODO: more description?

### Android

Android follows similar steps to iOS. We have to generate the code for Android, and then we have to write some native code to make it work.

#### Generate the Code - Android

To generate the code for Android, we need to manually invoke CodeGen. This is done similarly to what we did for iOS: first, we need to add the package to the app and then we need to invoke a script.

```sh title="Running CodeGen for Android"
cd MyApp
yarn add ../RTNCalculator
cd android
./gradlew generateCodegenArtifactsFromSchema --rerun-tasks
```

This script first adds the package to the app, in the same way iOS does. Then, after moving to the `android` folder, it invokes a Gradle task to create the generated code.

:::note
To run **CodeGen**, you need to enable the **New Architecture** in the Android app. This can be done by opening the `gradle.properties` files and by switching the `newArchEnabled` property from `false` to `true`.
:::

The generated code is stored in the `MyApp/node_modules/rtn-calculator/android/build/generated/source/codegen` folder and it has this structure:

> TODO: ordering issue. I can't actually get codegen to work at this point. Only works once I add the native files following. Probably similar issue for Fabric guide. Including generated file structure for later ease of use:

Reviewer: Do you have errors here or the CodeGen just does not generate anything?

Reviewer: We went further in the exploration and you were right. Android requires other files in order to be picked up by the Codegen. Specifically, it needs also:
-  The Android Manifest
- A file that implements the `ReactPackage` interface or that extends the `TurboReactPackage` class

Author: From the conversation in Discord, I'm not sure still exactly how much code I would write for that `TurboReactPackage` class before running Codegen; especially since I know I've run into issues on the Android side with creating an incomplete or incorrect interface before and then changing it.

import com.facebook.react.module.model.ReactModuleInfoProvider;
import com.facebook.react.TurboReactPackage;
import com.facebook.react.uimanager.ViewManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

public class CalculatorPackage extends TurboReactPackage {

  @Nullable
  @Override
  public NativeModule getModule(String name, ReactApplicationContext reactContext) {
      if (name.equals(CalculatorModule.NAME)) {
          return new CalculatorModule(reactContext);
      } else {
          return null;
      }
  }

  @Override
  public ReactModuleInfoProvider getReactModuleInfoProvider() {
      return () -> {
          final Map<String, ReactModuleInfo> moduleInfos = new HashMap<>();
          moduleInfos.put(
                  CalculatorModule.NAME,
                  new ReactModuleInfo(
                          CalculatorModule.NAME,
                          CalculatorModule.NAME,
                          false, // canOverrideExistingModule
                          false, // needsEagerInit
                          true, // hasConstants
                          false, // isCxxModule

>                           false // isTurboModule

Reviewer: I think this should be `true`. @cortinico what do you think?

Reviewer 2: Yup correct, this should be fixed

title: TurboModules
---

import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';

> If you've worked with React Native, you may be familiar with the concept of Native Modules, which allow JavaScript and platform-native code to communicate over the React Native "bridge", which handles cross-platform serialization via JSON.

Author: addressed

TurboModules are the next iteration on Native Modules that provide a few extra [benefits](./why):

- Strongly typed interfaces that are consistent across platforms

> - The ability to create shared C++ code for use across platforms

Author: @cortinico can you elaborate for me more what you want to convey by including "C++-only" here? Is there a misunderstanding you're trying to avoid? To my eye, I wouldn't assume the original wording means that I can share code in other languages as well.

(Side note: it would have to be written "C++-only", which looks a bit odd with the "++-", so I would try to reword this once I have a better understanding of intent.

> - The ability to create shared C++ code for use across platforms

Reviewer: I think we are saying two different things:
* In the original formulation, we wanted to convey that we can share code between iOS and Android using C++ code.
* In the suggestion by Nico, we are suggesting that we can create TM that can be used on both platforms and that contains only C++ code (this can be used for TM that do not need to access specific platform's features).

Both statements are true. I would specify both or look for a formulation that implies both meanings. What do you think?

> - The ability to create shared C++ code for use across platforms

Author: Tried to rephrase this to the best of my ability to explain both those ideas (thanks for laying them out so clearly @cipolleschi). Let me know what you think!

- Lazy loading of modules, allowing for faster app startup
- The use of JSI, a JavaScript interface for native code, which allows for more efficient communication between native and JavaScript code than the bridge

This guide will show you how to create a basic TurboModule.

:::caution
TurboModules only work with the **New Architecture** enabled.
To migrate to the **New Architecture**, follow the [Migration guide](../new-architecture-intro)
:::

## How to Create a TurboModule

To create a TurboModule, we need to:

> 1. Define a set of JavaScript specifications.

Author: One change: if it's no longer a "set", I would say "the JavaScript specification", not plural.

> 1. Define a set of JavaScript specifications.

Author: addressed

> 2. Configure the module and inspect the code created by Codegen.

Reviewer 2: ```suggestion
2. Configure the module and let the Codegen generate the scaffolding.
```

scaffolding can be replaced with boilerplate or other similar terms.

I would just not mention 'inspect the code' in the list of necessary steps.

Author: I like scaffolding, I think that's very clear about the role it plays, and implies you're expected to fill in more code.

I'm guessing the original intent in the "inspect the code" wording (from the Fabric Guide) was to talk about how we're going to read the scaffolding generated so we understand what native code to write, which does seem like a useful part of the process to convey, to me?

Author: Also just a note, but a few different suggestions left mention using "the CodeGen". Just to make sure we're clear, I don't think this is grammatically correct, it should either be "let CodeGen generate" or "let the CodeGen CLI tool (or process, or script, etc) generate..."

Author: Addressed, but leaving open for discussion if needed

Reviewer: > I'm guessing the original intent in the "inspect the code" wording (from the Fabric Guide) was to talk about how we're going to read the scaffolding generated so we understand what native code to write, which does seem like a useful part of the process to convey, to me?

That's correct, that was my intent there. The user will need to read the code generated by CodeGen in order to know how to use the generated types and files.

In this bullet list, I think that both the sentences fit. I will use the one that requires less effort from the reader (I think it is Nico's suggestion)

3. Write the native code to finish implementing the module.

## 1. Folder Setup

> In order to keep the module decoupled from the app, it's a good idea to define the module separately from the app, and then add it as a dependency to your app later.

Reviewer 2: nit: we can add it that this will make it easier to open-source it as a library afterwards if needed

Author: addressed

1. The file **must** be named `Native<MODULE_NAME>`, with a `.js` or `.jsx` extension when using Flow, or a `.ts`, or `.tsx` extension when using TypeScript. Codegen will only look for files matching this pattern.
2. The file must export a `TurboModuleRegistrySpec` object.

<Tabs groupId="turbomodule-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
<TabItem value="flow">

```typescript title="NativeCalculator.js"
// @flow
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}
export default (TurboModuleRegistry.get<Spec>(
  'RTNCalculator'
): ?Spec);
```

</TabItem>
<TabItem value="typescript">

```typescript title="NativeCalculator.ts"
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}

export default (TurboModuleRegistry.get<Spec>(
  'RTNCalculator'

> ) as Spec | null);

Reviewer 2: Why is this `| null` needed?

Author: This is my understanding of the return value of `TurboModuleRegistry.get` can be: either an object that implements `Spec` or null. Is that incorrect?

Author: I might not be understanding: this is the TypeScript example, but that looks like Flow syntax?

Reviewer: I agree with Lizzi here. For TypeScript examples is better to use TypeScript idiomatic syntax.

```

</TabItem>
</Tabs>

At the beginning of the spec files are the imports:

- The `TurboModule` type, which defines the base interface for all TurboModules
- The `TurboModuleRegistry` JavaScript module, which contains functions for loading TurboModules

The second section of the file contains the interface specification for the TurboModule. In this case, the interface defines the `add` function which takes two numbers and returns a promise that resolves to a number. This interface type **must** be named `Spec` for a TurboModule.

Finally, we invoke `TurboModuleRegistry.get`, passing the module's name, which will load the TurboModule if it's available. 

:::caution
We are writing JavaScript files importing types from libraries, without setting up a proper node module and installing its dependencies. Your IDE will not be able to resolve the import statements and you may see errors and warnings. This is expected and will not cause problems when you add the module to your app.

> :::

Reviewer 2: Nit: should we move the `package.json` above so this warning is not necessary. Then in the `Module Configuration` section we just focus on configuring the `codegenConfig`?

Author: > should we move the package.json above so this warning is not necessary.

that doesn't stop VSCode from generating warnings, for me.

```json title="package.json"
{
  "name": "rtn-calculator",
  "version": "0.0.1",
  "description": "Add numbers with TurboModules",
  "react-native": "js/index",
  "source": "js/index",
  "files": [
    "js",
    "android",
    "ios",
    "rtn-calculator.podspec",
    "!android/build",
    "!ios/build",
    "!**/__tests__",
    "!**/__fixtures__",
    "!**/__mocks__"
  ],
  "keywords": [
    "react-native",
    "ios",
    "android"
  ],
  "repository": "https://github.com/<your_github_handle>/rtn-calculator",
  "author": "<Your Name> <your_email@your_provider.com> (https://github.com/<your_github_handle>)",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/<your_github_handle>/rtn-calculator/issues"
  },
  "homepage": "https://github.com/<your_github_handle>/rtn-calculator#readme",
  "devDependencies": {},
  "peerDependencies": {
    "react": "*",
    "react-native": "*"
  },
  "codegenConfig": {
    "libraries": [
      {
        "name": "RTNCalculatorSpec",
        "type": "modules",
        "jsSrcsDir": "js"
      }
    ]
  }
}
```

> ### iOS: Create the `podspec` file

Author: Point of interest: [The style guide recommends putting Android before iOS](https://github.com/facebook/react-native-website/blob/main/STYLEGUIDE.md#android-before-ios), but I don't think that's been consistently followed. Is that still something we'd like to keep?

Reviewer 2: In this case, I would prefer local consistency over global consistency. IMHO let's keep the same ordering we followed in the other pages. We can re-sort them after we merged all of them 👍

end
```

The `.podspec` file has to be a sibling of the `package.json` file and its name is the one we set in the `package.json`'s `name` property: `rtn-calculator`.

The first part of the file prepares some variables we will use throughout the rest of it. Then, there is a section that contains some information used to configure the pod, like its name, version, and description. Finally, we have a set of dependencies that are required by the New Architecture.

### Android: Create the `build.gradle` file

Next, create a `build.gradle` file in the `android` folder, with the following contents:

```kotlin title="build.gradle"
buildscript {
  ext.safeExtGet = {prop, fallback ->
    rootProject.ext.has(prop) ? rootProject.ext.get(prop) : fallback
  }
  repositories {
    google()
    gradlePluginPortal()
  }
  dependencies {
    classpath("com.android.tools.build:gradle:7.1.1")
  }
}

apply plugin: 'com.android.library'
apply plugin: 'com.facebook.react'

android {
  compileSdkVersion safeExtGet('compileSdkVersion', 31)

  defaultConfig {
    minSdkVersion safeExtGet('minSdkVersion', 21)
    targetSdkVersion safeExtGet('targetSdkVersion', 31)

>     buildConfigField("boolean", "IS_NEW_ARCHITECTURE_ENABLED", "true")

Author: addressed

  }
}

repositories {
  maven {
    // All of React Native (JS, Obj-C sources, Android binaries) is installed from npm
    url "$projectDir/../node_modules/react-native/android"
  }
  mavenCentral()
  google()
}

dependencies {
  implementation 'com.facebook.react:react-native:+'
}

react {
    jsRootDir = file("../js/")
    libraryName = "RTNCalculator"

>     codegenJavaPackageName = "com.RTNCalculator"

Author: Yeah, I just matched the Fabric guide: https://github.com/facebook/react-native-website/blob/fd81aa82a7aceb9ff3b10d7b279a3de1055daf94/docs/the-new-architecture/pillars-fabric-components.md?plain=1#L281

>     codegenJavaPackageName = "com.RTNCalculator"

Reviewer 2: Yup let's keep it aligned then :+1:

}
```

Of interest in the `build.gradle` file:

> - The `defaultConfig` block, within the `android` block, adds a `buildConfigField` to enable the New Architecture.

Author: addressed

- The `react` block configures the CodeGen process. For Android, we need to specify:
  - the `jsRootDir`, which contains the relative path to the JavaScript specs
  - the `libraryName` we will use to link the library in the app.
  - the `codegenJavaPackageName` which corresponds to the name of the Java package we will use for the code generated by **CodeGen**.

## 4. Native Code

For the final step in getting your TurboModule ready to go, you'll need to write some native code to connect the JavaScript side to the native platforms. This process requires two main steps:

- Run **CodeGen** to see what it generates.
- Write your native code, implementing the generated interfaces.

> When developing a React Native app that uses a TurboMOdule, it is responsibility of the app to actually generate the code using **CodeGen**. However, when developing a TurboModule as a library, we need to reference the generated code, and it is therefore useful to see what the app will generate.

Author: addressed

As first step for both iOS and Android, this guide shows how to execute manually the scripts used by **CodeGen** to generate the required code. Further information on **CodeGen** can be found [here](/docs/pillars-codegen.md)

:::caution

> The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.

Author: This was mentioned in the Fabric guide: https://github.com/facebook/react-native-website/blob/fd81aa82a7aceb9ff3b10d7b279a3de1055daf94/docs/the-new-architecture/pillars-fabric-components.md?plain=1#L305. Should that also be changed? I'm not really familiar with what "distributing prebuilts" means in this specific context.

> The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.

Reviewer: When a distributor builds a library and distribute it to the public so that users don't have to build it themselves, they are creating a prefab.

We have ABI issues when a prefab uses a version of a library that is used also by the app and the version used by the prefab is different from the one used by the app (and potentially incompatible). For example, if a TM uses ReactNative 0.68 and we create and distribute a pre-built version of it and it is used by an app that runs 0.69, we may have issues.

> Should that also be changed?

Yes, I will update it there as well.

> The code generated by the **CodeGen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows to avoid any ABI incompatibility and to ensure that a consistent version of the **CodeGen** is used.

Author: Updated to be similar to the Fabric guide, but I made a slight edit:

```
The code generated by **Codegen** in this step should not be committed to the versioning system. React Native apps are able to generate the code when the app is built. This allows an app to ensure that all libraries have code generated for the correct version of React Native.
```

I'm not sure if this matches the intent; but I felt "ensure that a consistent version of the **CodeGen** is used across all the libraries." was a bit ambiguous -- all the libraries in what context? I assumed it meant all of a given app's libraries.

:::

### iOS

#### Generate the code - iOS

To run Codegen for the iOS platform, we need to open a terminal and run the following command:

```sh title="Running CodeGen for iOS"
cd MyApp
yarn add ../RTNCalculator

> cd ..

Author: > the extra ..

By this, do you mean in the `yarn add` command? That would be needed, yarn would need to be run from `MyApp`

As for the rest, I copied the way the Fabric guide did it.

> cd ..

Reviewer: We can remove the `cd ..` line, but in that case we have to update the next command as it follows:
```sh
node node_modules/react-native/scripts/generate-artifacts.js \
  --path . \
  --outputPath ../RTNCalculator/generated/
```
I think that the command is clearer if we run it from the parent folder, using `cd ..`, because it uses less `.` and `..` which adds more cognitive load for the reader.. What do you think?

> cd ..

Author: I also think it's a bit clearer from the parent. I'd rather be a bit more verbose for clarity's sake.

node MyApp/node_modules/react-native/scripts/generate-artifacts.js \
  --path MyApp/ \
  --outputPath RTNCalculator/generated/
```

This script first adds the `RTNCalculator` module to the app with `yarn add`. Then, it invokes Codegen via the `generate-artifacts.js` script.

> The `--path` option specifies the path to the app, while the `--outputPath` option tells the script where to output the generated code.

Author: addressed

The output of this process is the following folder structure:

```sh
generated
└── build
    └── generated
        └── ios
            ├── FBReactNativeSpec
            │   ├── FBReactNativeSpec-generated.mm
            │   └── FBReactNativeSpec.h
            ├── RCTThirdPartyFabricComponentsProvider.h
            ├── RCTThirdPartyFabricComponentsProvider.mm

>             ├── RTNCalculatorSpec

Author: Something that was surprising to me: after I generated this the first time, I realized I'd made a mistake in naming the module (RN instead of RTN), so changed the code and ran this again.

Instead of replacing the generated folder contents completely, it added the new `RTNCalculatorSpec` _alongside_ the `RNCalculatorSpec` that had been removed. If I totally wiped the generated files and re-generated them, `RNCalculatorSpec` was gone. Is this expected behavior?

Reviewer 2: > Instead of replacing the generated folder contents completely

IIRC the generate-artifacts script is not responsible of wiping the output folder, but just of reading the input and producing an output in the specified folder.

On Android: the cleanup is done by the Gradle Plugin. On iOS we should have a similar mechanism (cc @cipolleschi)

Reviewer: The cleanup mechanism on iOS is there on `main` but not in 0.68 nor in 0.69. :( 
At the moment, the folder has to be removed manually, unfortunately.

Author: Should the docs include a warning in case anyone runs into this?

Reviewer 2: > @cortinico, we wipe the `build` folder every time we run the CodeGen on Android?

Answering for future reference: yes we do wipe the generated folder every time the Codegen needs to run:

https://github.com/facebook/react-native/blob/8a33b75f55ba429e9e7a897179557aa8bb49ab35/packages/react-native-gradle-plugin/src/main/kotlin/com/facebook/react/tasks/GenerateCodegenSchemaTask.kt#L49-L52

This is to prevent stale files from being around.

> Should the docs include a warning in case anyone runs into this?

IMHO not. I would not mention this as this is how the whole Android/Gradle environment works (things are not re-executed unless needed). If there is a issue, is most likely a bug that needs to be addressed on our end 👍

            │   ├── RTNCalculatorSpec-generated.mm
            │   └── RTNCalculatorSpec.h
            └── react
                └── renderer
                    └── components
                        └── rncore
                            ├── ComponentDescriptors.h
                            ├── EventEmitters.cpp
                            ├── EventEmitters.h
                            ├── Props.cpp
                            ├── Props.h
                            ├── RCTComponentViewHelpers.h
                            ├── ShadowNodes.cpp
                            └── ShadowNodes.h
```

The relevant path for the TurboModule interface is `generated/build/generated/ios/RTNCalculatorSpec`.

See the [CodeGen](./pillars-codegen) section for further details on the generated files.

#### Write the Native iOS Code

Now add the Native code for your TurboModule. Create two files in the `RTNCalculator/ios` folder:

1. The `RTNCalculator.h`, a header file for the module.
2. The `RTNCalculator.mm`, the implementation of the module.

##### RTNCalculator.h

```objc title="RTNCalculator.h"
#import <React/RCTBridgeModule.h>

> @interface RTNCalculator : NSObject <RCTBridgeModule>

Author: The corresponding file in the Fabric guide also includes `NS_ASSUME_NONNULL_BEGIN` in this file, but the RNNewArch example doesn't. Wasn't sure if that was important to include?

Author: addressed

##### RTNCalculator.mm

```objc title="RTNCalculator.mm"
#import "RTNCalculatorSpec.h"
#import "RTNCalculator.h"

@implementation RTNCalculator

RCT_EXPORT_MODULE(RTNCalculator)

RCT_REMAP_METHOD(add, addA:(NSInteger)a
                      andB:(NSInteger)b
                withResolver:(RCTPromiseResolveBlock) resolve
                withRejecter:(RCTPromiseRejectBlock) reject)
{
    NSNumber *result = [[NSNumber alloc] initWithInteger:a+b];
    resolve(result);
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params
{
    return std::make_shared<facebook::react::NativeCalculatorSpecJSI>(params);
}

@end
```

The most important call is to the `RCT_EXPORT_MODULE`, which is required to export the module so that React Native can load the TurboModule.

Then the `RCT_REMAP_METHOD` macro is used to expose the `add` method.

:::info
There are other macros that can be used to export modules and methods. You view the code that specifies them [here](https://github.com/facebook/react-native/blob/main/React/Base/RCTBridgeModule.h).
:::

> TODO: more description?

Reviewer 2: Asking @cipolleschi here for a feedback

Reviewer: I would add a small paragraph on the `getTurbomodule:` method:
```
Finally, the `getTurboModule:` is used by the New Architecture to create a proper instance of the TurboModule so that the JavaScript side can properly invoke its methods. The function is defined in (and requested by) the `RTNCalculatorSpec.h` file that is generated by CodeGen.
```
Or something similar.

The final structure within the Android library should look like this:

```title="Android Folder Structure"
android
├── build.gradle
└── src
    └── main
        ├── AndroidManifest.xml
        └── java
            └── com
                └── RTNCalculator
                    ├── CalculatorModule.java
                    └── CalculatorPackage.java
```

##### AndroidManifest.xml

```xml title="AndroidManifest.xml"
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
          package="com.RTNCalculator">
</manifest>
```

This is a small manifest file that defines the package for our module.

##### CalculatorModule.java

```java title="CalculatorModule.java"
package com.RTNCalculator;

import androidx.annotation.NonNull;
import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import java.util.Map;
import java.util.HashMap;

public class CalculatorModule extends NativeCalculatorSpec {

> 

Reviewer 2: Let's add a:
```java
public static String NAME = "RTNCalculator";
```

and replace all the strings both here and inside `CalculatorPackage` with this static field.

import java.util.HashMap;
import java.util.Map;

public class CalculatorPackage extends TurboReactPackage {

  @Nullable
  @Override
  public NativeModule getModule(String name, ReactApplicationContext reactContext) {
      if (name.equals("RTNCalculator")) {
          return new CalculatorModule(reactContext);
      } else {
          return null;
      }
  }

  @Override
  public ReactModuleInfoProvider getReactModuleInfoProvider() {
      return () -> {
          final Map<String, ReactModuleInfo> moduleInfos = new HashMap<>();
          moduleInfos.put(
                  "RTNCalculator",
                  new ReactModuleInfo(
                          "RTNCalculator",
                          "RTNCalculator",
                          false, // canOverrideExistingModule
                          false, // needsEagerInit
                          true, // hasConstants
                          false, // isCxxModule
                          false // isTurboModule
          ));
          return moduleInfos;
      };
  }

}
```

This is the last piece of Native Code for Android. It defines the Package object that will be used by the app to load the module.

> ## 5. Adding the Fabric Component To Your App

Reviewer 2: ```suggestion
## 5. Adding the TurboModule To Your App
```

Author: addressed, thanks for catching that 🤦‍♀️

    1. Add the import for the calculator:
        ```diff
            #include <answersolver.h>
        +   #include <RTNCalculator.h>
        ```
    1. Add the following check in the `MainApplicationModuleProvider` constructor:
        ```diff
            // auto module = samplelibrary_ModuleProvider(moduleName, params);
            // if (module != nullptr) {
            //    return module;
            // }

        +    auto module = RTNCalculator_ModuleProvider(moduleName, params);
        +    if (module != nullptr) {
        +        return module;
        +    }

            return rncore_ModuleProvider(moduleName, params);
        }
        ```

### JavaScript

Now you can use your TurboModule calculator in your app!

Here's an example App.js file using the `add` method:

```js title="App.js"
/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 *
 * @format
 * @flow strict-local
 */
import React from 'react';
import {useState} from 'react';
import type {Node} from 'react';
import {SafeAreaView, StatusBar, Text, Button} from 'react-native';
import RTNCalculator from 'rtn-calculator/js/NativeCalculator.js';

const App: () => Node = () => {

>   const [currentResult, setResult] = useState<number | null>(null);

Reviewer 2: ```suggestion
  const [result, setResult] = useState<number | null>(null);
```

Author: Why this change, just to make it match the `[x, setX]` pattern? not a big deal, just will need to rename the `const` in the `onPress` to avoid conflict

Author: updated

Reviewer 2: > just to make it match the [x, setX] pattern?

Essentially yes. That's a nit, but the rest of docs we have that rely on `useState` follow this convention

Within `RTNCalculator`, create three subfolders: `js`, `ios`, and `android`.

The final result should look like this:

```sh
TurboModulesGuide
├── MyApp
└── RTNCalculator
    ├── android
    ├── ios
    └── js
```

## 2. JavaScript Specification

The **New Architecture** requires interfaces specified in a typed dialect of JavaScript (either [Flow](https://flow.org/) or [TypeScript](https://www.typescriptlang.org/)). **Codegen** will use these specifications to generate code in strongly-typed languages, including C++, Objective-C++, and Java.

There are two requirements the file containing this specification must meet:

1. The file **must** be named `Native<MODULE_NAME>`, with a `.js` or `.jsx` extension when using Flow, or a `.ts`, or `.tsx` extension when using TypeScript. Codegen will only look for files matching this pattern.
2. The file must export a `TurboModuleRegistrySpec` object.

<Tabs groupId="turbomodule-specs" defaultValue={constants.defaultJavaScriptSpecLanguages} values={constants.javaScriptSpecLanguages}>
<TabItem value="flow">

```typescript title="NativeCalculator.js"

> // @flow

Author: unsure if this should use `strict-local`? the RNNewArch example didn't, so I kept that

Reviewer 2: nope is file wiht just `// @flow`

import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}
export default (TurboModuleRegistry.get<Spec>(
  'RTNCalculator'
): ?Spec);
```

</TabItem>
<TabItem value="typescript">

```typescript title="NativeCalculator.ts"
import type { TurboModule } from 'react-native/Libraries/TurboModule/RCTExport';
import { TurboModuleRegistry } from 'react-native';

export interface Spec extends TurboModule {
  add(a: number, b: number): Promise<number>;
}

> export default (TurboModuleRegistry.get<Spec>(

Author: Just a point of curiosity: Does exporting default like this defeat lazy loading? Is lazy loading obtained a different way? Or is that lazy loading feature entirely on the native side? (I assume not or I'd expect `TurboModuleRegistry.get` to return a promise)

Author: From prior experience with things like this, my understanding is that `get` would be called as soon as this was imported in another file, so in the example app it's not being lazily loaded. We'd either have to export a function that calls `get` or conditionally `require` this export to have lazy loading, correct?

}

@end
```

The most important call is to the `RCT_EXPORT_MODULE`, which is required to export the module so that React Native can load the TurboModule.

Then the `RCT_REMAP_METHOD` macro is used to expose the `add` method.

Finally, the `getTurboModule` method gets an instance of the TurboModule so that the JavaScript side can invoke its methods. The function is defined in (and requested by) the `RTNCalculatorSpec.h` file that was generated earlier by Codegen.

:::info
There are other macros that can be used to export modules and methods. You view the code that specifies them [here](https://github.com/facebook/react-native/blob/main/React/Base/RCTBridgeModule.h).
:::

### Android

Android follows similar steps to iOS. We have to generate the code for Android, and then we have to write some native code to make it work.

#### Generate the Code - Android

To generate the code for Android, we need to manually invoke Codegen. This is done similarly to what we did for iOS: first, we need to add the package to the app and then we need to invoke a script.

```sh title="Running Codegen for Android"
cd MyApp
yarn add ../RTNCalculator
cd android

> ./gradlew generateCodegenArtifactsFromSchema --rerun-tasks

Reviewer 2: @cipolleschi Similarly for the Fabric guide, I believe we should not suggest to `--rerun-tasks`.
